### ISSUE TITLE
"Frame size exceeds frame max" (RabbitMQ 3.4.2)

### ISSUE THREAD

**Issue by atombender** [squaremo/amqp.node#135]
This happened suddenly. How do I debug this?

```
/tmp/node_modules/amqplib/lib/frame.js:55
      throw new Error('Frame size exceeds frame max');
            ^
Error: Frame size exceeds frame max
    at parseFrame (/tmp/node_modules/amqplib/lib/frame.js:55:13)
    at Connection.C.recvFrame (/tmp/node_modules/amqplib/lib/connection.js:592:15)
    at Socket.go (/tmp/node_modules/amqplib/lib/connection.js:465:30)
    at Socket.emit (events.js:92:17)
    at emitReadable_ (_stream_readable.js:426:10)
    at emitReadable (_stream_readable.js:422:5)
    at readableAddChunk (_stream_readable.js:165:9)
    at Socket.Readable.push (_stream_readable.js:127:10)
    at TCP.onread (net.js:528:21)
```

Adding some output shows that the negotiated frame size is 4096, and that the frame returned when it blows up is 4249. RabbitMQ reports nothing wrong.

**Comment by atombender** [comment]
For what it's worth, it only happens to me for messages that have bounced back from a dead-letter-exchange queue.

My setup: I have a main queue that has a DLX setting that shuffles nacks into a separate queue. That queue has a ttl setting and a DLX setting pointing back to the main queue. Thus, failed messages will be retried by the main queue after timing out.

My setup logic is as follows, in case you want to try replicating it:

```javascript
    Promise.all([
      channel.assertQueue(queueName, {
        durable: true,
        deadLetterExchange: exchangeName,
        deadLetterRoutingKey: formatRoutingKey({event: failureEventName})
      }),
      channel.assertQueue(failureQueueName, {
        durable: true,
        messageTtl: RETRY_DELAY,
        deadLetterExchange: exchangeName,
        deadLetterRoutingKey: formatRoutingKey({event: retryEventName})
      }),
      channel.bindQueue(queueName, exchangeName, formatRoutingKey(routingOptions)),
      channel.bindQueue(queueName, exchangeName, formatRoutingKey({event: retryEventName})),
      channel.bindQueue(failureQueueName, exchangeName, formatRoutingKey({event: failureEventName})),
      channel.prefetch(8)
    ]).then(...)
```

(`formatRoutingKey()` just generates a routing key according to my own conventions.)

**Comment by squaremo** [comment]
I'm pretty sure RabbitMQ will re-fragment messages to fit the particular connection, so it would unusual for it to send a too-big frame. What is the nature of the messages being sent through?

**Comment by atombender** [comment]
You'll need to be a little more specific about what you mean. Some of my messages are probably larger than 4K, though, if that helps. It seems to be happening quite randomly, and so far I have only seen the problem when the messages end up going via the DLX.

I also did some tracing, and before it fails, it reads around 29KB from the server in the `recvFrame` method. I can add more tracing if you can give me some pointers as to where to look.

**Comment by atombender** [comment]
Any ideas? This is happening constantly, and we have a system we'd like to put into production — we can't, now, as working around the protocol issue means purging the queue.

**Comment by squaremo** [comment]
> What is the nature of the messages being sent through?

The first thing I'd suspect is that the properties or headers of a message (rather than the content) encode to great than 4k, since AMQP has no provision for fragmenting anything other than message content. Is this possible in your application?

I'm not sure what RabbitMQ does in this situation, but it basically has no choice but to put an oversized frame on the wire (or drop the message).

You could raise the frame max -- it's set by default at the minimum (the minimum maximum), and increasing it won't have much of an effect other than letting through those frames through. To do that append e.g., '?frameMax=8192' to the connection URL.

> I also did some tracing, and before it fails, it reads around 29KB from the server in the recvFrame method

That's pretty normal I would think -- in usual operation it will read all it can from the socket, then parse that into frames.

**Comment by atombender** [comment]
We don't put much in the header, definitely not more than 4KB.

In case it helps, [here](https://gist.github.com/atombender/08fde20f595e7fb30f48]) is a gist with output from `rabbitmqadmin get ...`.

I can try increasing the frame limit.

**Comment by squaremo** [comment]
OK, with a hat-tip to @rade I can make an informed guess about what's going on. The trick is here https://www.rabbitmq.com/dlx.html:

> The dead-lettering process adds an array to the header of each dead-lettered message named x-death. This array contains an entry for each time the message was dead-lettered. Each such entry is a table that consists of several fields [...]

In other words, every time a message is bounced through the DLX it picks up another chunk of header, and I suspect this eventually gets to a point where RabbitMQ cannot encode the frame in less than frame_max and has to put it on the wire anyway.

If this is the case, there is no particular frame_max that will avoid the problem, although it may defer it -- in the absence of a fix in RabbitMQ (apparently there's a bug for it), you could enforce a limit on the number of bounces you do before moving the message somewhere else.

**Comment by atombender** [comment]
Good catch, thanks! I'll add logic to look at `x-death` and stop the loop past a certain time. Feel free to close this.

**Comment by squaremo** [comment]
Cheers -- hope you got it working OK.

**Comment by atombender** [comment]
The endless `x-death` header problem has finally been [fixed in RabbitMQ 3.5.1](https://github.com/rabbitmq/rabbitmq-server/releases/tag/rabbitmq_v3_5_1). The list will not grow if the queue name and rejection reason is the same as the previous entry.

**Comment by michaelklishin** [comment]
@atombender and there was a fix for that [in 3.5.2](https://github.com/rabbitmq/rabbitmq-server/releases/tag/rabbitmq_v3_5_2).
